test/json_unittest.cc

```diff
@@ -0,0 +1,385 @@
+    EXPECT_EQ(arr->size(), 1);
+    EXPECT_EQ((*arr)[0]->As<json::Number>()->value(), 2.0);
+  });
+}
+
+TEST(TrivialToStringTest, ArrayOneElement) {
+  GuardJsonException([](){
+    EXPECT_TO_STRING(json::array({json::number(2.0)}), "[2.0]");
+  });
+}
+
+
+// Trival Dict
+
+TEST(TrivialParseTest, DictEmpty) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("{}");
+    EXPECT_TRUE(v->As<json::Object>()->empty());
+  });
+}
+
+TEST(TrivialToStringTest, DictEmpty) {
+  GuardJsonException([](){
+    EXPECT_TO_STRING(json::object({}), "{}");
+  });
+}
+
+TEST(TrivialParseTest, DictOneField) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("{\"f\":2}");
+    const auto& obj = v->As<json::Object>();
+    EXPECT_EQ(obj->size(), 1);
+    EXPECT_EQ(obj->Get<json::Number>("f"), 2.0);
+  });
+}
+
+TEST(TrivialToStringTest, DictOneField) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::object({
+      {"f", json::number(2.0)}
+    });
+    EXPECT_TO_STRING(v, "{\"f\":2.0}");
+  });
+}
+
+
+// String edge cases
+
+/*
+ * TODO(igorp): Seems that yajl does not support valid JSON.
+TEST(EdgeTest, StringWithNewLine) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"\n\"");
+    EXPECT_EQ(v->As<json::String>()->value(), "\n");
+    // Output should be escaped with \n as that is more canonical.
+    EXPECT_TO_STRING(v, "\"\n\"");
+  });
+}
+*/
+
+TEST(EdgeTest, StringWithNonUnicodeEscape) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"foo\\nbar\"");
+    EXPECT_EQ(v->As<json::String>()->value(), "foo\nbar");
+    // Output should be escaped as that is more canonical.
+    EXPECT_TO_STRING(v, "\"foo\\nbar\"");
+  });
+}
+
+TEST(EdgeTest, StringWithEveryEscape) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"x\\\\x\\/x\\bx\\fx\\nx\\rx\\tx\"");
+    EXPECT_EQ(v->As<json::String>()->value(), "x\\x/x\bx\fx\nx\rx\tx");
+    // Output should be escaped as that is more canonical.
+    // Since / does not need to be escaped, output does not escape that.
+    EXPECT_TO_STRING(v, "\"x\\\\x/x\\bx\\fx\\nx\\rx\\tx\"");
+  });
+}
+
+TEST(EdgeTest, StringWithUnicodeEscape) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"foo\\u000abar\"");
+    //EXPECT_EQ(v->As<json::String>()->value(), "foo\nbar");
+    // Output should be escaped as that is more canonical.
+    EXPECT_TO_STRING(v, "\"foo\\nbar\"");
+  });
+}
+
+TEST(EdgeTest, StringWithUTF8) {
+  GuardJsonException([](){
+    // This is Korean for "Hello World!".
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"안녕 세상아!\"");
+    EXPECT_EQ(v->As<json::String>()->value(), "안녕 세상아!");
+    EXPECT_TO_STRING(v, "\"안녕 세상아!\"");
+  });
+}
+
+
+// Number edge cases
+
+TEST(EdgeTest, PositiveNumbers) {
```

Comment:
Is this a parse/ToString test? Would it make sense to also test JSON construction:
```cpp
    json::value v = json::array({
      0, 0e+0, 0e-0, 0e0, 0E+0, 0E-0, 0E0,
      0.0, 0.0e+0, 0.0e-0, 0.0e0, 0.0E+0, 0.0E-0, 0.0E0,
      10, 10e+0, 10e-0, 10e0, 10E+0, 10E-0, 10E0,
      10.0, 10.0e+0, 10.0e-0, 10.0e0, 10.0E+0, 10.0E-0, 10.0E0,
      123, 123.456, 789, 1.05, 1.999e-99,
    });
```
?

Reply:
That's testing the C++ parser though.  I could have split it into two parts, but since we do a lot of that already (and therefore ensure good coverage) I figured an end-to-end test is enough here.

Reply:
This is also testing the contract provided by those functions and constructors... Let's do this in the `RealisticConstruction` tests.